src/lsm/scan_groove.zig

```diff
@@ -0,0 +1,892 @@
+                .field = prefix,
+                .timestamp = timestamp,
+            });
+        }
+
+        /// Initializes a Scan over the secondary index specified by `index`,
+        /// searching for an exact match in the `CompositeKey`'s prefix.
+        /// Produces the criteria equivalent to `WHERE field = $value`.
+        /// Results are ordered by `timestamp`.
+        pub fn scan_prefix(
+            self: *Self,
+            comptime index: std.meta.FieldEnum(Groove.IndexTrees),
+            buffer: *const ScanBuffer,
+            snapshot: u64,
+            value: CompositeKeyPrefix(index),
+            timestamp_range: TimestampRange,
+            direction: Direction,
+        ) *Scan {
+            const field = comptime index_to_dispatcher(index);
+            const ScanImpl = ScanImplType(field);
+            return scan_add(
+                self,
+                field,
+                ScanImpl.init(
+                    self.tree(index),
+                    buffer,
+                    snapshot,
+                    key_from_value(index, value, timestamp_range.min),
+                    key_from_value(index, value, timestamp_range.max),
+                    direction,
+                ),
+            ) catch unreachable; //TODO: define error handling for the query API.
+        }
+
+        /// Initializes a Scan over the secondary index specified by `index`,
+        /// searching for an inclusive range of `CompositeKey`s.
+        /// Results are ordered by `prefix`, but not by `timestamp`.
+        /// Range scans cannot be used in the `merge_{union,intersection,difference}` functions.
+        pub fn scan_range(
+            self: *Self,
+            comptime index: std.meta.FieldEnum(Groove.IndexTrees),
+            buffer: *const ScanBuffer,
+            snapshot: u64,
+            min: CompositeKeyType(index),
+            max: CompositeKeyType(index),
+            direction: Direction,
+        ) *Scan {
+            const field = comptime index_to_dispatcher(index);
+            const ScanImpl = ScanImplType(field);
+            return scan_add(
+                self,
+                field,
+                ScanImpl.init(
+                    self.tree(index),
+                    buffer,
+                    snapshot,
+                    min.key_from_value(),
+                    max.key_from_value(),
+                    direction,
+                ),
+            ) catch unreachable; //TODO: define error handling for the query API.
+        }
+
+        /// Initializes a Scan performing the union operation over multiple scans.
+        /// E.g. S₁ ∪ S₂ ∪ Sₙ.
+        /// Produces the criteria equivalent to
+        /// `WHERE <condition_1> OR <condition_2> OR <condition_N>`.
+        ///
+        /// All scans must yield results in the same direction.
+        pub fn merge_union(
+            self: *Self,
+            scans: []const *Scan,
+            direction: Direction,
+        ) *Scan {
```

Comment:
Good point.

- Instead of a bitset, I implemented it as a field `assigned : bool`.
- I'm also checking if the scan hasn't been executed before (`state != idle`).
- Made some code reorganization and moved the merge logic to `scan_merge.zig`.

In order to access the state, now `ScanTree` and `ScanMerge{Union|...}` shares the same state enum tag.
As a bonus tip, I added one more state to track which scan needs to read, so I could remove the `TODO` notes and the `next_tick` workaround! https://github.com/tigerbeetle/tigerbeetle/pull/1054#discussion_r1409509166